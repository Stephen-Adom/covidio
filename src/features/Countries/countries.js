/* eslint-disable react/require-default-props */
import { Column } from 'primereact/column';
import { DataTable } from 'primereact/datatable';
import PropTypes from 'prop-types';
import React from 'react';
import { useParams } from 'react-router';

const Countries = ({ countries }) => {
  const { continent } = useParams();
  const columns = [
    { field: 'Country', header: 'Country' },
    { field: 'TotalCases', header: 'Total Cases' },
    { field: 'NewCases', header: 'New Cases' },
    { field: 'Critical', header: 'Critical' },
    { field: 'ActiveCases', header: 'Active Cases' },
    { field: 'TotalDeaths', header: 'Total Deaths' },
  ];

  const specific = countries.map((country) => ({
    Country: country.country,
    TotalCases: country.cases,
    NewCases: country.todayCases,
    Critical: country.critical,
    ActiveCases: country.active,
    TotalDeaths: country.deaths,
  }));

  const dynamicCols = columns.map((col) => (
    <Column key={col.field} field={col.field} header={col.header} sortable />
  ));

  return (
    <section className="p-2">
      <div className="mb-5">
        <div className=" px-2">
          {specific && (
            <DataTable
              value={specific}
              responsiveLayout="scroll"
              header={continent ? `Countries in ${continent}` : 'All Countries'}
              showGridlines
              className="data-table shadow shadow-slate-500"
            >
              {dynamicCols}
            </DataTable>
          )}
        </div>
      </div>
    </section>
  );
};

Countries.propTypes = {
  countries: PropTypes.shape([]),
};

export default Countries;
