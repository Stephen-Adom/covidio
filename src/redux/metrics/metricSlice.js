import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import {
  fetchAllCountries,
  fetchContinent,
  fetchGlobalData,
  fetchAllContinents,
} from '../api/endpoints';

const initialState = {
  global: {},
  countries: [],
  continent: {},
  allContinents: [],
  status: '',
  error: '',
};

export const globalInfo = createAsyncThunk('metrics/all', async (thunkAPI) => {
  try {
    const response = await fetchGlobalData();
    return response.data;
  } catch (err) {
    return thunkAPI.rejectWithValue(err.response.data);
  }
});

export const getAllContinents = createAsyncThunk('metrics/all-continents', async (thunkAPI) => {
  try {
    const response = await fetchAllContinents();
    return response.data;
  } catch (err) {
    return thunkAPI.rejectWithValue(err.response.data);
  }
});

export const getAllCountries = createAsyncThunk(
  'metrics/countries',
  async (thunkAPI) => {
    try {
      const response = await fetchAllCountries();
      return response.data;
    } catch (err) {
      return thunkAPI.rejectWithValue(err.response.data);
    }
  },
);

export const getContinent = createAsyncThunk(
  'metrics/continent',
  async (payload, thunkAPI) => {
    try {
      const response = await fetchContinent(payload);
      return response.data;
    } catch (err) {
      return thunkAPI.rejectWithValue(err.response.data);
    }
  },
);

const metricSlice = createSlice({
  name: 'metrics',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(globalInfo.pending, (state) => {
        const IsPending = state;
        IsPending.status = 'pending';
      })
      .addCase(globalInfo.fulfilled, (state, action) => {
        const IsFulfilled = state;
        IsFulfilled.status = 'fulfilled';
        IsFulfilled.global = action.payload;
        IsFulfilled.error = '';
      })
      .addCase(globalInfo.rejected, (state, action) => {
        const IsRejected = state;
        IsRejected.status = 'rejected';
        IsRejected.global = {};
        IsRejected.error = action.payload;
      })

      .addCase(getAllContinents.pending, (state) => {
        const IsPending = state;
        IsPending.status = 'pending';
      })

      .addCase(getAllContinents.fulfilled, (state, action) => {
        const IsFulfilled = state;
        IsFulfilled.status = 'fulfilled';
        IsFulfilled.allContinents = action.payload;
        IsFulfilled.error = '';
      })

      .addCase(getAllContinents.rejected, (state, action) => {
        const IsRejected = state;
        IsRejected.status = 'rejected';
        IsRejected.error = action.payload;
      })

      .addCase(getAllCountries.pending, (state) => {
        const IsPending = state;
        IsPending.status = 'pending';
      })
      .addCase(getAllCountries.fulfilled, (state, action) => {
        const IsFulfilled = state;
        IsFulfilled.status = 'fulfilled';
        IsFulfilled.countries = action.payload;
        IsFulfilled.error = '';
      })
      .addCase(getAllCountries.rejected, (state, action) => {
        const IsRejected = state;
        IsRejected.status = 'rejected';
        IsRejected.countries = [];
        IsRejected.error = action.payload;
      })
      .addCase(getContinent.pending, (state) => {
        const IsPending = state;
        IsPending.status = 'pending';
      })
      .addCase(getContinent.fulfilled, (state, action) => {
        const IsFulfilled = state;
        IsFulfilled.status = 'fulfilled';
        IsFulfilled.continent = action.payload;
        IsFulfilled.error = '';
      })
      .addCase(getContinent.rejected, (state, action) => {
        const IsRejected = state;
        IsRejected.status = 'rejected';
        IsRejected.continent = [];
        IsRejected.error = action.payload;
      });
  },
});

export default metricSlice.reducer;
