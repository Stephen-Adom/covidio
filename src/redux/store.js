import { configureStore } from '@reduxjs/toolkit';
import metricSlice from './metrics/metricSlice';

const store = configureStore({
  reducer: {
    metrics: metricSlice,
  },
});

export default store;
