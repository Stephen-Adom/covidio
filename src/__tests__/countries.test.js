import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { RouterProvider } from 'react-router-dom';
import Countries from '../features/Countries/countries';
import store from '../redux/store';
import router from '../route/routes';
import { getAllCountries, getAllContinents } from '../redux/metrics/metricSlice';

describe('countries & continents', () => {
  it('should fetch all countries and save in state', async () => {
    await store.dispatch(getAllCountries()).then(async (response) => {
      expect(response.payload.length).toEqual(store.getState().metrics.countries.length);
    });
  });

  it('should fetch all continents and save in state', async () => {
    await store.dispatch(getAllContinents()).then(async (response) => {
      expect(response.payload.length).toEqual(store.getState().metrics.allContinents.length);
    });
  });

  it('should render list of countries', async () => {
    await store.dispatch(getAllCountries()).then(async () => {
      const tree = renderer.create(
        <Provider store={store}>
          <RouterProvider router={router}>
            <Countries type="country" />
          </RouterProvider>
        </Provider>,
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  it('should render list of continents', async () => {
    await store.dispatch(getAllContinents()).then(async () => {
      const tree = renderer.create(
        <Provider store={store}>
          <RouterProvider router={router}>
            <Countries type="continent" />
          </RouterProvider>
        </Provider>,
      ).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
