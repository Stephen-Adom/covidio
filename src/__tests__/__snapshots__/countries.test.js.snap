// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`countries & continents should render list of continents 1`] = `
<div
  className="App"
>
  <section
    className="splash-container"
  >
    <div
      className="relative logo-container"
    >
      <img
        alt="covido"
        className="bounce-in-fwd"
        src="covido-logo.png"
        width="100%"
      />
      <h1
        className="logo tracking-in-expand"
      >
        Covidio
      </h1>
    </div>
  </section>
</div>
`;

exports[`countries & continents should render list of countries 1`] = `
<div
  className="App"
>
  <section
    className="splash-container"
  >
    <div
      className="relative logo-container"
    >
      <img
        alt="covido"
        className="bounce-in-fwd"
        src="covido-logo.png"
        width="100%"
      />
      <h1
        className="logo tracking-in-expand"
      >
        Covidio
      </h1>
    </div>
  </section>
</div>
`;
